ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,165 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+findWhereQ :: GenericQ [HsLocalBinds GhcTc]
+findWhereQ = everything (<>) $ mkQ [] (pure . findWhere)
+  where
+    findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> HsLocalBinds GhcTc
+    findWhere = grhssLocalBinds
+
+-- | Find all bindings for **one** where clause.
+findBindingsQ :: GenericQ (Maybe WhereBindings)
+findBindingsQ = something (mkQ Nothing findBindings)
+  where
+    findBindings :: NHsValBindsLR GhcTc -> Maybe WhereBindings
+    findBindings (NValBinds binds sigs) =
+      Just $ WhereBindings
+        { bindings = concat $ mapMaybe (something (mkQ Nothing findBindingIds) . snd) binds
+        , existingSigNames = concatMap findSigIds sigs
+        }
+
+    findBindingIds :: LHsBindLR GhcTc GhcTc -> Maybe [WhereBinding]
+    findBindingIds bind = case unLoc bind of
+      FunBind{..} ->
+        let whereBinding = WhereBinding (unLoc fun_id) (getLoc fun_id)
+                           (col (getLoc fun_id) - col (getLoc bind))
+        in Just $ pure whereBinding
+      PatBind{..} -> Just $ (everything (<>) $ mkQ [] (fmap (uncurry wb) . maybeToList . findIdFromPat)) pat_lhs
+        where
+          wb id srcSpan = WhereBinding id srcSpan (col srcSpan - col (getLoc pat_lhs))
+      _           -> Nothing
+      where
+        col = srcSpanStartCol . realSrcSpan
+
+    -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
+    findIdFromPat :: Pat GhcTc -> Maybe (Id, SrcSpan)
+    findIdFromPat (VarPat _ located) = Just (unLoc located, getLoc located)
+    findIdFromPat _                  = Nothing
+
+    findSigIds :: GenLocated l (Sig GhcRn) -> [IdP GhcRn]
+    findSigIds (L _ (TypeSig _ names _)) = map unLoc names
+    findSigIds _                         = []
+
+-- | Provide code lens for where bindings.
+whereClauseInlayHints :: PluginMethodHandler IdeState Method_TextDocumentInlayHint
+whereClauseInlayHints state plId (InlayHintParams _ (TextDocumentIdentifier uri) visibleRange)  = do
+  enabled <- liftIO $ runAction "inlayHint.config" state $ usePropertyAction #whereInlayHintOn plId properties
+  if not enabled then pure $ InL [] else do
+    nfp <- getNormalizedFilePathE uri
+    (WhereBindingTypeSigsResult (localBindings, sigMap), pm)
+        <- runActionE "InlayHint.GetWhereBindingTypeSigs" state $ useWithStaleE GetWhereBindingTypeSigs nfp
+    let bindingToInlayHints id sig = generateWhereInlayHints (T.pack $ printName (idName id)) (maybe "_" T.pack sig)
+
+        -- | Note there may multi ids for one binding,
+        -- like @(a, b) = (42, True)@, there are `a` and `b`
+        -- in one binding.
+        inlayHints =
+          [ bindingToInlayHints bindingId bindingSig bindingRange offset
+          | WhereBindings{..} <- localBindings
+          , let sigSpans = getSrcSpan <$> existingSigNames
+          , WhereBinding{..} <- bindings
+          , let bindingSpan = getSrcSpan (idName bindingId)
+          , let bindingSig = Map.lookup bindingId sigMap
+          , bindingSpan `notElem` sigSpans
```

Comment:
This seems a little complicated? I wonder if we could just track the `Id`s that have signatures already?

Reply:
It seems that `Sig GhcTc` does not exists (not sure), that is to say, `Id` (`IdP GhcTc`) does not exist in `HsLocalBinds`, and currently only `Name` (`IdP GhcRn`) can be obtained from `Sig GhcRn`.
ref: https://hackage.haskell.org/package/ghc-9.10.1/docs/GHC-Hs-Binds.html#v:NValBinds

On the other hand, the pass of `LHsBinds` is `GhcTc`, and the `Name` obtained from `bindingId` is different from the`Name` in `Sig` because their `Unique` is different.
ref: https://hackage.haskell.org/package/ghc-9.10.1/docs/GHC-Types-Unique.html#v:nonDetCmpUnique

If the `Id`s that have signatures is not easily available, I've tried comparing via `OccName`, but maybe `SrcSpan` is still better?